xbox/xbox.c

```diff
@@ -0,0 +1,1155 @@
+    break;
+  case GL_SRC1_RGB:
+    t->src_rgb[1] = param;
+    break;
+  case GL_SRC2_RGB:
+    t->src_rgb[2] = param;
+    break;
+
+  case GL_SRC0_ALPHA:
+    t->src_alpha[0] = param;
+    break;
+
+  case GL_OPERAND0_RGB:
+    t->src_operand_rgb[0] = param;
+    break;
+
+  case GL_OPERAND1_RGB:
+    t->src_operand_rgb[1] = param;
+    break;
+
+  case GL_OPERAND2_RGB:
+    t->src_operand_rgb[2] = param;
+    break;
+
+  case GL_OPERAND0_ALPHA:
+    t->src_operand_alpha[0] = param;
+    break;
+
+  default:
+    unimplemented("%d", pname);
+    assert(false);
+    return;
+  }
+}
+
+
+// Pixel readback
+GL_API void GL_APIENTRY glReadPixels (GLint x, GLint y, GLsizei width, GLsizei height, GLenum format, GLenum type, void *pixels) {
+  // Not implemented, only used in screenshots
+  unimplemented();
+}
+
+
+// Lighting
+GL_API void GL_APIENTRY glLightModelf (GLenum pname, GLfloat param) {
+  switch(pname) {
+  default:
+    unimplemented("%d", pname);
+    assert(false);
+    return;
+  }
+}
+
+GL_API void GL_APIENTRY glLightModelfv (GLenum pname, const GLfloat *params) {
+  switch(pname) {
+  default:
+    unimplemented("%d", pname);
+    assert(false);
+    return;
+  }
+}
+
+GL_API void GL_APIENTRY glLightfv (GLenum light, GLenum pname, const GLfloat *params) {
+  assert(light >= GL_LIGHT0);
+  unsigned int light_index = light - GL_LIGHT0;
+  assert(light_index < 4); //FIXME: Not sure how many lights Xbox has; there's probably some constant we can use
+  switch(pname) {
+  default:
+    unimplemented("%d", pname);
+    assert(false);
+    return;
+  }
+}
+
+
+// Materials
+GL_API void GL_APIENTRY glMaterialfv (GLenum face, GLenum pname, const GLfloat *params) {
+
+  if (face == GL_FRONT_AND_BACK) {
+    glMaterialfv(GL_FRONT, pname, params);
+    glMaterialfv(GL_BACK, pname, params);
+    return;
+  }
+
+  assert(face == GL_FRONT || face == GL_BACK);
+  switch(pname) {
+  unimplemented(); //FIXME: Missing from XGU
+  default:
+    unimplemented("%d", pname);
+    assert(false);
+    return;
+  }
+}
+
+
+// Pixel pushing
+GL_API void GL_APIENTRY glPixelStorei (GLenum pname, GLint param) {
+  switch(pname) {
+  case GL_PACK_ALIGNMENT:
+    assert(param == 1);
+    break;
+  case GL_UNPACK_ALIGNMENT:
+    assert(param == 1);
+    break;
+  default:
+    unimplemented("%d", pname);
+    assert(false);
+    break;
+  }
+}
+
+
+
+
+//FIXME: Use stuff like `ptr = MmAllocateContiguousMemoryEx(size, 0, 0x3ffb000, 0, 0x404);` to alloc buffer / tex
+
+// Initialization
+
+#include <hal/video.h>
+#include <hal/xbox.h>
+#include <pbkit/pbkit.h>
+#include <xboxkrnl/xboxkrnl.h>
+
+__attribute__((destructor)) static void exit_xbox(void) {
+  assert(false);
+}
+
+static HAL_SHUTDOWN_REGISTRATION shutdown_registration;
+static NTAPI VOID shutdown_notification_routine (PHAL_SHUTDOWN_REGISTRATION ShutdownRegistration) {
+  exit_xbox();
+}
+
+__attribute__((constructor)) static void setup_xbox(void) {
+  //FIXME: Remap log, see OpenSWE1R
+
```

Comment:
Getting issues with `debugPrint` in XQEMU. Should allocate the XVideo framebuffer (I did this in another project, but I don't remember which one).